File: solution/113/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        int now = 0;
        vector<vector<int> > result;
        vector<int> path;
        getPath(root, now, sum, result, path);
        return result;
    }
    void getPath(TreeNode* root,int &now, int sum, vector<vector<int> >&result, vector<int>&path){
        if(!root)return;
        now += root->val;
        path.push_back(root->val);
        if(!root->left&&!root->right&&now==sum)result.push_back(path);
        getPath(root->left, now, sum, result, path);
        getPath(root->right, now, sum, result, path);
        now -= root->val;
        path.pop_back();
    }
};

File: solution/113/solution.py
import copy
class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        result = []
        path = []
        now = 0
        self.getpath(root, now, sum, path, result)
        return result

    def getpath(self, root, now, sum, path, result):
        if not root:return
        now += root.val
        path.append(root.val)
        if not root.left and not root.right and sum == now:result.append(copy.copy(path))
        self.getpath(root.left, now, sum, path, result)
        self.getpath(root.right, now, sum, path, result)
        now -= root.val
        path.pop()

